doc/fgpu.design.rst

```diff
@@ -597,147 +597,96 @@ channels) when
+
+The first simplification we make is to pre-compute the weights with the mixer
+(on the CPU). Let :math:`z_k = w_k e^{2\pi j\omega k}`. Then the equation
+becomes
+
+.. math::
+
+   y_{b+i} = e^{2\pi j\omega S(b+i)} \sum_{k=0}^{T-1} x_{S(b+i)+k} z_k.
+
+For now we'll focus on just the summation, and deal with the exponential later.
+For simplicity, assume :math:`T` is a multiple of :math:`S` (although the
+implementation does not require it) and let :math:`W = \frac{T}{S}`. Then we
+can write :math:`j` as :math:`pS + q` and rewrite this equation as
+
+.. math::
+
+   y_{b+i} = e^{2\pi j\omega S(b+i)} \sum_{p=0}^{W - 1} \sum_{q=0}^{S-1} x_{S(b+i+p) + q} z_{pS + q}.
+
+The kernel iterates first over :math:`q`, then :math:`p`, then :math:`i`. A
+separate accumulator variable is kept for each value of :math:`i`.
+For a given :math:`q`, we only need :math:`C + W - 1` different values
+of :math:`x` (since that's the range of :math:`i+p`). We decode them all into
+an array before iterating over :math:`q` and :math:`i` to update the
+accumulators.
+
+When :math:`q` is advanced, we need to decode a new set of :math:`C + W - 1`
+samples. These immediately follow the previous set. We take advantage of this:
+in many cases, the new sample occupies (at least partially) the same 32-bit
+word from which we obtained the previous sample. By keeping those
+:math:`C + W - 1` words around, we get a head-start on decoding the new
+sample.
+
+Choosing :math:`C` is a trade-off. Larger values of :math:`C` clearly increase
+register pressure. However they reduce the number of loads required: each work
+item decodes :math:`(C + W - 1)S` samples to produce :math:`C` outputs, which
+is an average of :math:`S + \frac{(W - 1)S}{C}`.
+To further reduce the global memory traffic, all the samples and weights are
+copied to local memory at the start of the kernel. The results are also first
+transposed in local memory before being written back to global memory, to
+improve the global memory access pattern.
+
+The implementation relies heavily on loop unrolling. Provided that :math:`CS`
+samples occupy a whole number of 32-bit words (so that different work-items are
+loading samples from the same bit positions within a word), all the conditional
+logic involved in decoding the samples can be evaluated at compile-time.
 
 Mixer signal
 ~~~~~~~~~~~~
 Care needs to be taken with the precision of the argument to the mixer signal.
-Simply evaluating the sine and cosine of :math:`2\pi f t` when
+Simply evaluating the sine and cosine of :math:`2\pi \omega t` when
 :math:`t` is large can lead to a catastrophic loss of precision, as the
-product :math:`f t` will have a large integer part and leave few bits for
-the fractional part. Even passing :math:`f` in single precision can lead
+product :math:`\omega t` will have a large integer part and leave few bits for
+the fractional part. Even passing :math:`\omega` in single precision can lead
```

Comment:
Again, gauging this straight swap from `f` to `\omega`, was **f** being correctly used in the original discussion?
Or am I (more likely) still missing some Narrowband 101 here?

Reply:
No, it was incorrect previously. I'm slightly embarrassed to have missed it last time around.